sdk/order-planner-common/src/context/ExecutionStorageMapper.ts

```diff
@@ -0,0 +1,103 @@
+import { isValueReference, steps } from '@summerfi/sdk-common/simulation'
+import { Maybe } from '@summerfi/sdk-common/common'
+import {
+  ParamsMapping as InputSlotsMapping,
+  StorageInputsMapType,
+  StorageOutputsMapType,
+} from '../interfaces/Types'
+import assert from 'assert'
+import { BaseAction } from '../actions/BaseAction'
+
+export class ExecutionStorageMapper {
+  private slotsMap: Map<string, number> = new Map()
+
+  // Slot number starts from 1 because the contract will subtract 1 from the slot number
+  private currentSlot: number = 1
+
+  public addStorageMap<Step extends steps.Steps, Action extends BaseAction>(params: {
+    step: Step
+    action: Action
+    connectedInputs: Partial<StorageInputsMapType<Step, Action>>
+    connectedOutputs: Partial<StorageOutputsMapType<Step, Action>>
+  }): InputSlotsMapping {
+    const baseSlot = this.currentSlot
+    const stepOutpus = (params.step.outputs as unknown) ?? {}
+
+    for (const stepOutputName of Object.keys(stepOutpus)) {
+      if (params.connectedOutputs !== undefined) {
+        const actionOutputName =
+          params.connectedOutputs[stepOutputName as keyof StorageOutputsMapType<Step, Action>]
+
+        if (actionOutputName === undefined) {
+          continue
+        }
+
+        const slotKey = this._getSlotKey({
+          stepName: params.step.name,
+          referenceName: stepOutputName,
+        })
+
+        const slotOffset = params.action.config.storageOutputs.indexOf(actionOutputName)
+        assert(slotOffset !== -1, 'Output not found in action storage outputs')
+
+        this.slotsMap.set(slotKey, baseSlot + slotOffset)
+      }
+
+      this.currentSlot++
+    }
+
+    return this._resolveParamsMapping(params)
+  }
+
+  public getOutputSlot(params: {
+    stepName: string
+    referenceName: string | number | symbol
+  }): Maybe<number> {
+    const key = this._getSlotKey(params)
+    return this.slotsMap.get(key)
+  }
+
+  private _getSlotKey(params: {
+    stepName: string
+    referenceName: string | number | symbol
+  }): string {
+    return `${params.stepName}-${String(params.referenceName)}`
+  }
+
+  private _resolveParamsMapping<Step extends steps.Steps, Action extends BaseAction>(params: {
+    action: Action
+    step: Step
+    connectedInputs: Partial<StorageInputsMapType<Step, Action>>
+  }): InputSlotsMapping {
+    const paramsMapping: InputSlotsMapping = [0, 0, 0, 0]
```

Comment:
Well, yes and no. There is an tacit agreement between the actions and the caller that there are not more than 4 params mapping, but there is no enforcement for it. The actions are not checking the input length of the mapping array, so it could actually be less or more. I just followed suit to not overcomplicate things, but I think we could expand it if needed, and even send less data, but I was afraid of changing such a low level feature